c/tskit/trees.c

```diff
@@ -6803,3 +6803,268 @@ tsk_treeseq_divergence_matrix(const tsk_treeseq_t *self, tsk_size_t num_samples,
+    edge_list_t **head, edge_list_t **tail, tsk_blkalloc_t *heap, tsk_id_t edge)
+{
+    int ret = 0;
+    edge_list_t *x = NULL;
+
+    x = tsk_blkalloc_get(heap, sizeof(*x));
+    if (x == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    x->edge = edge;
+    x->extended = false;
+    x->next = NULL;
+
+    if (*tail == NULL) {
+        *head = x;
+    } else {
+        (*tail)->next = x;
+    }
+    *tail = x;
+out:
+    return ret;
+}
+
+static void
+remove_unextended(edge_list_t **head, edge_list_t **tail)
+{
+    edge_list_t *px, *x;
+
+    px = *head;
+    while (px != NULL && !px->extended) {
+        px = px->next;
+    }
+    *head = px;
+    if (px != NULL) {
+        px->extended = false;
+        x = px->next;
+        while (x != NULL) {
+            if (x->extended) {
+                x->extended = false;
+                px->next = x;
+                px = x;
+            }
+            x = x->next;
+        }
+    }
+    *tail = px;
+}
+
+static int
+do_extend(tsk_treeseq_t *self, int direction)
+{
+    // Note: this modifies the edge table, but it does this by (a) removing
+    // some edges, and (b) extending left/right endpoints of others,
+    // while keeping order the same, and so this maintains sortedness
+    // (so, there is no need to sort afterwards).
+    int ret = 0;
+    tsk_id_t *degree;
+    tsk_id_t tj, ret_id;
+    tsk_id_t e, e1, e2, e_in;
+    double *near_side, *far_side;
+    tsk_blkalloc_t edge_list_heap;
+    edge_list_t *edges_in_head, *edges_in_tail;
+    edge_list_t *edges_out_head, *edges_out_tail;
+    edge_list_t *ex1, *ex2, *ex_in;
+    tsk_edge_table_t edges;
+    tsk_edge_t edge;
+    double there, left, right;
+    tsk_id_t sign_int;
+    bool forwards = (direction == TSK_DIR_FORWARD);
+    tsk_tree_position_t tree_pos;
+    bool valid;
+
+    // need to do this so tsk_safe_free works on it if it's not initialized
+    memset(&edges, 0, sizeof(edges));
+
+    degree = tsk_malloc(self->tables->nodes.num_rows * sizeof(*degree));
+    if (degree == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(degree, 0x00, self->tables->nodes.num_rows * sizeof(*degree));
+
+    ret = tsk_blkalloc_init(&edge_list_heap, 8192);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_tree_position_init(&tree_pos, self, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_edge_table_copy(&self->tables->edges, &edges, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_edge_table_clear(&self->tables->edges);
+    if (ret != 0) {
+        goto out;
+    }
+
+    if (forwards) {
+        sign_int = 1;
+        near_side = edges.left;
+        far_side = edges.right;
+    } else {
+        sign_int = -1;
+        near_side = edges.right;
+        far_side = edges.left;
+    }
+    edges_in_head = NULL;
+    edges_in_tail = NULL;
+    edges_out_head = NULL;
+    edges_out_tail = NULL;
+
+    if (forwards) {
```

Comment:
@duncanMR what do you think about adding something like ``tsk_tree_position_step(&tree_pos, direction)`` for this? That would make it quite neat for things like this.

Reply:
That would be a useful thing to have in the API and trivial to implement. I'd be happy to do it